src/event_listener.rs

```diff
@@ -42,26 +58,49 @@ impl<E: EntityEvent> On<E> {
     /// systems is that the callback system can access a resource with event data,
     /// [`ListenerInput`]. You can more easily access this with the system params
     /// [`Listener`](crate::callbacks::Listener) and [`ListenerMut`](crate::callbacks::ListenerMut).
-    pub fn run<Marker>(callback: impl IntoSystem<(), (), Marker>) -> Self {
-        Self {
-            phantom: PhantomData,
-            callback: CallbackSystem::New(Box::new(IntoSystem::into_system(callback))),
+    pub fn run<Marker>(&mut self, callback: impl IntoSystem<(), (), Marker>) -> ListenerHandle<E> {
+        let handle = ListenerHandle::new(); // You need to implement this method
+        self.callbacks.push((handle.clone(), CallbackSystem::New(Box::new(IntoSystem::into_system(callback)))));
+        handle
+    }
+
+    /// Create an empty event listener.
+    pub fn init_callbacks(&mut self) {
+        self.callbacks = Vec::new();
+    }
+
+    /// Remove a callback from this event listener.
+    pub fn remove(&mut self, handle: ListenerHandle<E>)
+    where
+        ListenerHandle<E>: PartialEq,
+    {
+        self.callbacks.retain(|(h, _)| *h != handle);
+    }
+
+    /// Replace a callback with a new one. This is useful for hot-reloading systems.
+    pub fn replace<Marker>(&mut self, handle: ListenerHandle<E>, callback: impl IntoSystem<(), (), Marker>)
+    where
+        ListenerHandle<E>: PartialEq,
+    {
+        if let Some((_, cb)) = self.callbacks.iter_mut().find(|(h, _)| *h == handle) {
+            *cb = CallbackSystem::New(Box::new(IntoSystem::into_system(callback)));
         }
     }
 
     /// Add a single [`Command`] any time this event listener is triggered. The command must
     /// implement `From<E>`.
-    pub fn add_command<C: From<ListenerInput<E>> + Command + Send + Sync + 'static>() -> Self {
-        Self::run(|event: Res<ListenerInput<E>>, mut commands: Commands| {
+    pub fn add_command<C: From<ListenerInput<E>> + Command + Send + Sync + 'static>(&mut self) -> ListenerHandle<E> {
+        self.run(|event: Res<ListenerInput<E>>, mut commands: Commands| {
             commands.add(C::from(event.to_owned()));
         })
     }
 
     /// Get mutable access to [`Commands`] any time this event listener is triggered.
     pub fn commands_mut(
+        &mut self,
         mut func: impl 'static + Send + Sync + FnMut(&mut ListenerInput<E>, &mut Commands),
-    ) -> Self {
-        Self::run(
+    ) -> ListenerHandle<E> {
```

Comment:
We should revert the changes to the method signatures from here down. Rather than mutating self, create a new empty event listener, mutate and then return it.